ui/appui-react/src/appui-react/widget-panels/ToolSettings.tsx

```diff
@@ -87,42 +84,26 @@ export function useHorizontalToolSettingNodes() {
     InternalFrontstageManager.activeToolSettingsProvider
       ?.horizontalToolSettingNodes
   );
+  const [emptySettings] = React.useState(() => [EmptyToolSettingsEntry()]);
   React.useEffect(() => {
-    const handleToolActivatedEvent = () => {
+    return UiFramework.frontstages.onToolActivatedEvent.addListener(() => {
       const nodes =
         InternalFrontstageManager.activeToolSettingsProvider
           ?.horizontalToolSettingNodes;
-      if (!nodes || nodes.length === 0) setSettings([EmptyToolSettingsEntry()]);
-      else setSettings(nodes);
-    };
-    UiFramework.frontstages.onToolActivatedEvent.addListener(
-      handleToolActivatedEvent
-    );
-    return () => {
-      UiFramework.frontstages.onToolActivatedEvent.removeListener(
-        handleToolActivatedEvent
-      );
-    };
+      setSettings(nodes);
+    });
   }, [setSettings]);
 
   React.useEffect(() => {
-    const handleToolSettingsReloadEvent = () => {
+    return UiFramework.frontstages.onToolSettingsReloadEvent.addListener(() => {
       const nodes =
         InternalFrontstageManager.activeToolSettingsProvider
           ?.horizontalToolSettingNodes;
-      if (!nodes || nodes.length === 0) setSettings([EmptyToolSettingsEntry()]);
-      else setSettings(nodes);
-    };
-    UiFramework.frontstages.onToolSettingsReloadEvent.addListener(
-      handleToolSettingsReloadEvent
-    );
-    return () => {
-      UiFramework.frontstages.onToolSettingsReloadEvent.removeListener(
-        handleToolSettingsReloadEvent
-      );
-    };
+      setSettings(nodes);
+    });
```

Comment:
Not really, having two separate `useEffect` calls makes it easier to clean up the listeners. You could still reuse the handler, but it is fairly straightforward in this case.

Reply:
Since you use lodash, you can `map` event handlers and `pipe` clean up functions in a single `useEffect` like so

```
React.useEffect(() => pipe(...[
  UiFramework.frontstages.onToolActivatedEvent, 
  UiFramework.frontstages.onToolSettingsReloadEvent
].map(e => e.addListener(() => {
  const nodes = InternalFrontstageManager.activeToolSettingsProvider
    ?.horizontalToolSettingNodes;
   setSettings(nodes);  
})), [setSettings]);
```